### PR TITLE
Handle running out of shapes in `Object#dup`

### PR DESCRIPTION
**Pull request by casperisfine**
There is a handful of call sites where we may transition to OBJ_TOO_COMPLEX_SHAPE if we just ran out of shapes, but that weren't handling it properly.

cc @tenderlove @jemmaissroff

### CODE REVIEW

File: test/ruby/test_shapes.rb [hunk 1]
```diff
@@ -226,6 +226,34 @@ def initialize
...
+      o = Object.new
+      i = 0
+      while RubyVM::Shape.shapes_available > 1
+        o.instance_variable_set(:"@i#{i}", 1)
+        i += 1
+      end
+
+      GC.disable # FIXME: if GC is enabled we run into another bug
+
+      a.dup
```
Comment (casperisfine) [inline test/ruby/test_shapes.rb:253]: So here we create a first instance of `A` that wasn't right sized and had to spill over in a malloc.

Then after running out of shapes we dup it, Ruby tries to rebuild the correct shape tree and run out of shapes in the middle of the process.

File: test/ruby/test_shapes.rb [hunk 2]
```diff
@@ -226,6 +226,34 @@ def initialize
...
+      a = A.new
+
+      o = Object.new
+      i = 0
+      while RubyVM::Shape.shapes_available > 1
+        o.instance_variable_set(:"@i#{i}", 1)
+        i += 1
+      end
+
+      GC.disable # FIXME: if GC is enabled we run into another bug
```
Comment (casperisfine) [inline test/ruby/test_shapes.rb:251]: I'll dig into that one but the failure is weird:

```
ruby/miniruby(sigsegv) (null):0
/usr/lib/system/libsystem_platform.dylib(_sigtramp+0x38) [0x19d246a24]
ruby/miniruby(mark_tbl_no_pin+0x0) [0x104799a6c] gc.c:7336
ruby/miniruby(gc_mark_children) gc.c:7336
ruby/miniruby(gc_mark_stacked_objects+0x78) [0x1047a1a0c] gc.c:7435
ruby/miniruby(gc_mark_stacked_objects_incremental) gc.c:7467
ruby/miniruby(gc_marks_step+0x8) [0x10479ee20] gc.c:8672
ruby/miniruby(gc_marks_continue) gc.c:8692
ruby/miniruby(gc_continue) gc.c:2480
ruby/miniruby(newobj_alloc+0xb8) [0x10479e864] gc.c:2807
ruby/miniruby(newobj_of0+0x90) [0x1047911bc] gc.c:2919
ruby/miniruby(newobj_of) gc.c:2936
ruby/miniruby(rb_method_entry_alloc+0x20) [0x104927cf8] ./vm_method.c:650
ruby/miniruby(rb_method_entry_complement_defined_class) ./vm_method.c:709
ruby/miniruby(prepare_callable_method_entry+0x13c) [0x1049292b4] ./vm_method.c:1245
ruby/miniruby(callable_method_entry_or_negative+0x188) [0x104928cf0] ./vm_method.c:1363
ruby/miniruby(callable_method_entry+0x10) [0x104939d70] ./vm_method.c:1389
ruby/miniruby(rb_callable_method_entry) ./vm_method.c:1396
ruby/miniruby(gccct_method_search_slowpath) ./vm_eval.c:452
ruby/miniruby(rb_funcallv_scope+0x38) [0x10492a090] ./vm_eval.c:1059
ruby/miniruby(rb_funcallv) ./vm_eval.c:1080
```

Comment (casperisfine) [reply test/ruby/test_shapes.rb]: Nvm, I figured it out.

**Review by peterzhu2118 — APPROVED** [review]
Looks good to me!

File: test/ruby/test_shapes.rb [hunk 3]
```diff
@@ -226,6 +226,52 @@ def initialize
...
+      while RubyVM::Shape.shapes_available > 0
+        o.instance_variable_set(:"@i#{i}", 1)
+        i += 1
+      end
+
+      tc = TooComplex.new
+      tc.instance_variable_set(:@a, 1)
+      tc.instance_variable_set(:@b, 2)
+    end;
+  end
```
Comment (casperisfine) [inline test/ruby/test_shapes.rb:247]: I found another crash on our CI, so I pushed a second commit to handle it.

I'm running yet another build to see if there's other cases we don't handle.

**Review by jemmaissroff — APPROVED** [review]

**Comment by casperisfine** [comment]
Yet another error spotted on our CI:

```
[BUG] Segmentation fault at 0x0000000000000008

---
-- Control frame information -----------------------------------------------
c:0112 p:---- s:0764 e:000763 CFUNC  :instance_variable_set
c:0111 p:0009 s:0758 e:000757 METHOD /tmp/bundle/ruby/3.3.0+0/gems/identity_cache-1.4.1/lib/identity_cache/cached/recursive/association.rb:45
c:0110 p:0015 s:0752 e:000751 METHOD /tmp/bundle/ruby/3.3.0+0/gems/identity_cache-1.4.1/lib/identity_cache/cached/recursive/association.rb:50
c:0109 p:0094 s:0746 e:000745 METHOD /tmp/bundle/ruby/3.3.0+0/gems/identity_cache-1.4.1/lib/identity_cache/cached/recursive/association.rb:35
c:0108 p:0005 s:0738 e:000737 BLOCK  /tmp/bundle/ruby/3.3.0+0/gems/identity_cache-1.4.1/lib/identity_cache/cached/recursive/association.rb:19
---

/usr/local/ruby/bin/ruby(rb_print_backtrace+0x14) [0x55d034be6487] /tmp/ruby-build/ruby-3.3.0-6e7b7eac484dc680a1698ae98d5d041de5fd1d6d/vm_dump.c:812
/usr/local/ruby/bin/ruby(rb_vm_bugreport) /tmp/ruby-build/ruby-3.3.0-6e7b7eac484dc680a1698ae98d5d041de5fd1d6d/vm_dump.c:1143
/usr/local/ruby/bin/ruby(rb_bug_for_fatal_signal+0x10e) [0x55d034d8b56e] /tmp/ruby-build/ruby-3.3.0-6e7b7eac484dc680a1698ae98d5d041de5fd1d6d/error.c:1065
/usr/local/ruby/bin/ruby(sigsegv+0x4d) [0x55d034b38b2d] /tmp/ruby-build/ruby-3.3.0-6e7b7eac484dc680a1698ae98d5d041de5fd1d6d/signal.c:920
/lib/x86_64-linux-gnu/libc.so.6(0x7fe8bc533520) [0x7fe8bc533520]
/usr/local/ruby/bin/ruby(do_hash+0xe) [0x55d034b45019] /tmp/ruby-build/ruby-3.3.0-6e7b7eac484dc680a1698ae98d5d041de5fd1d6d/st.c:320
/usr/local/ruby/bin/ruby(rb_st_insert) /tmp/ruby-build/ruby-3.3.0-6e7b7eac484dc680a1698ae98d5d041de5fd1d6d/st.c:1112
/usr/local/ruby/bin/ruby(RB_IMMEDIATE_P+0x0) [0x55d034ba424e] /tmp/ruby-build/ruby-3.3.0-6e7b7eac484dc680a1698ae98d5d041de5fd1d6d/variable.c:1431
/usr/local/ruby/bin/ruby(RB_SPECIAL_CONST_P) ./include/ruby/internal/special_consts.h:329
/usr/local/ruby/bin/ruby(rb_obj_written) ./include/ruby/internal/gc.h:803
/usr/local/ruby/bin/ruby(rb_complex_ivar_set) /tmp/ruby-build/ruby-3.3.0-6e7b7eac484dc680a1698ae98d5d041de5fd1d6d/variable.c:1432
/usr/local/ruby/bin/ruby(rb_obj_ivar_set+0x179) [0x55d034ba83e9] /tmp/ruby-build/ruby-3.3.0-6e7b7eac484dc680a1698ae98d5d041de5fd1d6d/variable.c:1636
/usr/local/ruby/bin/ruby(ivar_set+0xe) [0x55d034bac70e] /tmp/ruby-build/ruby-3.3.0-6e7b7eac484dc680a1698ae98d5d041de5fd1d6d/variable.c:1738
/usr/local/ruby/bin/ruby(rb_ivar_set) /tmp/ruby-build/ruby-3.3.0-6e7b7eac484dc680a1698ae98d5d041de5fd1d6d/variable.c:1757
/usr/local/ruby/bin/ruby(rb_obj_ivar_set_m+0x4b) [0x55d034a8251b] /tmp/ruby-build/ruby-3.3.0-6e7b7eac484dc680a1698ae98d5d041de5fd1d6d/object.c:2908
/usr/local/ruby/bin/ruby(vm_call_cfunc_with_frame_+0x10b) [0x55d034bc444b] /tmp/ruby-build/ruby-3.3.0-6e7b7eac484dc680a1698ae98d5d041de5fd1d6d/vm_insnhelper.c:3477
/usr/local/ruby/bin/ruby(vm_sendish+0x16a) [0x55d034bbcbfa] /tmp/ruby-build/ruby-3.3.0-6e7b7eac484dc680a1698ae98d5d041de5fd1d6d/vm_insnhelper.c:5555
/usr/local/ruby/bin/ruby(vm_exec_core+0xa5) [0x55d034bd62e5] /tmp/ruby-build/ruby-3.3.0-6e7b7eac484dc680a1698ae98d5d041de5fd1d6d/insns.def:822
```

I'll try to find a repro.

**Comment by casperisfine** [comment]
Hum, got one failure on the `ubuntu-i686` that seem legit:

```
1) Failure:
  TestShapes#test_run_out_of_shape_generic_ivar_set [/home/runner/work/ruby/ruby/src/test/ruby/test_shapes.rb:230]:
  pid 166379 killed by SIGABRT (signal 6) (core dumped)
  | -:19: [BUG] Expected shape_id entry in global iv table
  | ruby 3.3.0dev (2023-10-30T13:55:53Z pull/8799/merge 32138d1c94) [i686-linux-gnu]
  | 
  | -- Control frame information -----------------------------------------------
  | c:0003 p:---- s:0015 e:000014 CFUNC  :instance_variable_set
  | c:0002 p:0104 s:0009 E:001414 EVAL   -:19 [FINISH]
  | c:0001 p:0000 s:0003 E:001b48 DUMMY  [FINISH]
  | 
  | -- Ruby level backtrace information ----------------------------------------
  | -:19:in `<main>'
  | -:19:in `instance_variable_set'
  | 
  | -- Threading information ---------------------------------------------------
  | Total ractor count: 1
  | Ruby thread count for this ractor: 1
  | 
  | -- C level backtrace information -------------------------------------------
  | /home/runner/work/ruby/ruby/build/ruby(rb_print_backtrace+0x14) [0x5681493a] ../src/vm_dump.c:812
  | /home/runner/work/ruby/ruby/build/ruby(rb_vm_bugreport) ../src/vm_dump.c:1143
  | /home/runner/work/ruby/ruby/build/ruby(bug_report_end+0x0) [0x56905810] ../src/error.c:[1042](https://github.com/ruby/ruby/actions/runs/6693821642/job/18189422320?pr=8799#step:8:1048)
  | /home/runner/work/ruby/ruby/build/ruby(rb_bug_without_die) ../src/error.c:1042
  | /home/runner/work/ruby/ruby/build/ruby(die+0x0) [0x566262e8] ../src/error.c:[1050](https://github.com/ruby/ruby/actions/runs/6693821642/job/18189422320?pr=8799#step:8:1056)
  | /home/runner/work/ruby/ruby/build/ruby(rb_bug) ../src/error.c:[1052](https://github.com/ruby/ruby/actions/runs/6693821642/job/18189422320?pr=8799#step:8:1058)
  | /home/runner/work/ruby/ruby/build/ruby(rb_multi_ractor_p+0x0) [0x5661908b] ../src/variable.c:1693
  | /home/runner/work/ruby/ruby/build/ruby(rb_vm_lock_enter) ../src/vm_sync.h:74
  | /home/runner/work/ruby/ruby/build/ruby(rb_shape_set_shape_id) ../src/variable.c:1685
  | /home/runner/work/ruby/ruby/build/ruby(rb_shape_set_shape_id) ../src/variable.c:1665
  | /home/runner/work/ruby/ruby/build/ruby(rb_shape_set_too_complex+0x2f) [0x5676736f] ../src/shape.c:875
  | /home/runner/work/ruby/ruby/build/ruby(rb_evict_ivars_to_hash+0x67) [0x567d7637] ../src/variable.c:1446
  | /home/runner/work/ruby/ruby/build/ruby(generic_ivar_set+0x35) [0x567dcfa2] ../src/variable.c:1498
  | /home/runner/work/ruby/ruby/build/ruby(ivar_set) ../src/variable.c:1748
  | /home/runner/work/ruby/ruby/build/ruby(rb_ivar_set+0x65) [0x567dd055] ../src/variable.c:1757
  | /home/runner/work/ruby/ruby/build/ruby(rb_obj_ivar_set_m+0x4f) [0x566b975f] ../src/object.c:2908
  | /home/runner/work/ruby/ruby/build/ruby(vm_call_cfunc_with_frame_+0xce) [0x567eaf0e] ../src/vm_insnhelper.c:3477
  | /home/runner/work/ruby/ruby/build/ruby(vm_call_method_each_type+0x7e) [0x567f806e] ../src/vm_insnhelper.c:4391
  | /home/runner/work/ruby/ruby/build/ruby(vm_call_method+0x116) [0x567f8b76] ../src/vm_insnhelper.c:4543
  | /home/runner/work/ruby/ruby/build/ruby(vm_exec_core+0x4d3) [0x56805203] ../src/vm_insnhelper.c:5555
  | /home/runner/work/ruby/ruby/build/ruby(rb_vm_exec+0x99) [0x567f7189] ../src/vm.c:2472
  | /home/runner/work/ruby/ruby/build/ruby(rb_ec_exec_node+0xa8) [0x56627ee8] ../src/eval.c:287
  | /home/runner/work/ruby/ruby/build/ruby(ruby_run_node+0xb6) [0x5662cba6] ../src/eval.c:328
  | /home/runner/work/ruby/ruby/build/ruby(rb_main+0x21) [0x566267f6] ../src/main.c:39
  | /home/runner/work/ruby/ruby/build/ruby(main) ../src/main.c:58
  | /lib/i386-linux-gnu/libc.so.6(__libc_start_main+0xf5) [0xf7c2aed5]
```

**Pull request merged**